// ottstream.dataaccess/src/models/index.js
module.exports.Token = require('./user/token.model');
module.exports.User = require('./user/user.model');
module.exports.UserActivity = require('./user/user_activity.model');
module.exports.Balance = require('./payment/balance.model');
module.exports.Credit = require('./payment/credit.model');
module.exports.File = require('./file/file.model');
module.exports.Transaction = require('./payment/transaction.model');
module.exports.Invoice = require('./payment/invoice.model');
module.exports.History = require('./history/history.model');
module.exports.Credit = require('./payment/credit.model');
module.exports.CreditCard = require('./payment/credit_card/credit_card.model');
module.exports.PaymentGateway = require('./payment/payment_gateway.model');
module.exports.PaymentImplementation = require('./payment/payment_implementation.model');
module.exports.PaymentMethod = require('./payment/payment_method.model');
module.exports.ShippingProvider = require('./payment/shipping_provider.model');
module.exports.PermissionMethod = require('./role/permission.model');
module.exports.LanguageUnit = require('./language/language_unit.model');
module.exports.LanguageUnitTranslation = require('./language/language_unit_translation.model');
module.exports.Language = require('./language/language.model');
module.exports.Country = require('./country/country.model');
module.exports.Permission = require('./role/permission.model');
module.exports.Role = require('./role/role.model');
module.exports.IconType = require('./icon_type/icon_type.model');
module.exports.SystemVariable = require('./system/system_variables.model');
module.exports.BankName = require('./payment/bank_name.model');

module.exports.OttProvider = require('./ottprovider/ottprovider.model');
module.exports.OttProviderPermission = require('./ottprovider/ottprovider_permission.model');
module.exports.OttProviderAddress = require('./ottprovider/ottprovider_address.model');
module.exports.OttProviderEmail = require('./ottprovider/ottprovider_email.model');
module.exports.OttProviderPhone = require('./ottprovider/ottprovider_phone.model');
module.exports.OttProviderShippingProvider = require('./ottprovider/ottprovider_shipping_provider.model');
module.exports.OttProviderConversationProvider = require('./ottprovider/ottprovider_conversation_provider.model');
module.exports.OttProviderPaymentGateway = require('./ottprovider/ottprovider_payment_gateway.model');
module.exports.OttProviderOtherApi = require('./ottprovider/ottprovider_other_api.model');
module.exports.OttProviderPrinter = require('./ottprovider/ottprovider_printer.model');
module.exports.OttProviderUi = require('./ottprovider/ottprovider_ui.model');
module.exports.OttProviderInfo = require('./ottprovider/ottprovider_info.model');
module.exports.OttProviderPaymentMethod = require('./ottprovider/ottprovider_payment_method.model');
module.exports.OttProviderCategoryChannel = require('./ottprovider/ottprovider_category_channel.model');
module.exports.OttProviderInvoice = require('./ottprovider/ottprovider_invoice.model');
module.exports.Channel = require('./channel/channel.model');
module.exports.ChannelIconSet = require('./channel/channel_icon_set.model');
module.exports.ChannelIconSetType = require('./channel/channel_icon_set_type.model');
module.exports.ChannelCategory = require('./channel/channel_category.model');

module.exports.Package = require('./package/package.model');
module.exports.PackageOption = require('./package/package_options.model');
module.exports.PackageChannel = require('./package/package_channel.model');
module.exports.Subscription = require('./subscription/subsciption.model');

module.exports.PriceGroup = require('./price/price_group.model');
module.exports.Booking = require('./booking/booking.model');
module.exports.AgeGroup = require('./client/age_group.model');
module.exports.Currency = require('./currency/currency.model');
module.exports.CurrencyCountry = require('./currency/currency_country.model');

module.exports.ProductType = require('./product/product_type.model');
module.exports.Product = require('./product/product.model');

module.exports.Discount = require('./discount/discount.model');
module.exports.Client = require('./client/client.model');
module.exports.ClientActivity = require('./client/client_activity.model');
module.exports.ClientLocation = require('./client/client_location.model');
module.exports.ClientProfile = require('./client/client_profile.model');
module.exports.ClientPackage = require('./client/client_package.model');
module.exports.ClientUsedDevice = require('./client/client_used_device.model');
module.exports.ClientUsedDeviceActivity = require('./client/client_used_device_activity.model');
module.exports.ClientPaymentMethod = require('./client/client_payment_method.model');
module.exports.ClientBill = require('./client_bill/client_bill.model');
module.exports.Server = require('./server/server.model');
module.exports.Group = require('./group/group.model');
module.exports.Icon = require('./icon/icon.model');
module.exports.Equipment = require('./equipment/equipment.model');
module.exports.EquipmentInstaller = require('./equipment/equipment_installer.model');
module.exports.EquipmentType = require('./equipment/equipment_type.model');
module.exports.EquipmentSubscription = require('./subscription/equipment_subsciption.model');
module.exports.CalendarEvent = require('./calendar/calendar_event.model');

module.exports.DeviceOption = require('./device/device_options.model');
module.exports.Backup = require('./backup/backup.model');
module.exports.Comment = require('./comment/comment.model');
module.exports.Notification = require('./notification/notification.model');
module.exports.Geoip = require('./geoip/geoip.model');
module.exports.Shipping = require('./shipping/shipping.model');
module.exports.EasyshipCourier = require('./shipping/easyship_courier.model');
module.exports.Chat = require('./chat/chat.model');
module.exports.Sms = require('./sms/sms.model');
module.exports.Help = require('./help/help.model');
module.exports.TelegramBot = require('./telegram_bot/telegram_bot.model');
